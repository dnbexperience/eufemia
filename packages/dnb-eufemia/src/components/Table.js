/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index table to autogenerate all the components and extensions
 * Used by "prepareTables"
 */

import Table from './table/Table'
export * from './table/Table'
export default Table
