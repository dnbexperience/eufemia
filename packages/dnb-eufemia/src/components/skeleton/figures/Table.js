/**
 * Web Skeleton Figure
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import Context from '../../../shared/Context'
import classnames from 'classnames'

export default class SkeletonTable extends React.PureComponent {
  static contextType = Context
  static propTypes = {
    rows: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    children: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
      PropTypes.node
    ])
  }

  static defaultProps = {
    rows: 3,
    children: null
  }

  constructor(props) {
    super(props)

    const { rows } = props

    // Do this so we get the same result each time
    // because of static generated markup
    const fill = [70, 80, 60, 40, 50, 20, 0]
    this.rowsLength = new Array(rows).fill(true).map((_, i) => {
      const c = i % fill.length
      if (c === fill.length - 1) {
        fill.concat(fill.reverse())
      }
      return fill[c]
    })
  }

  render() {
    const {
      rows, // eslint-disable-line
      children,
      ...rest
    } = this.props

    return (
      <div
        className={classnames(
          'dnb-skeleton__figure',
          'dnb-skeleton__figure--show'
        )}
        aria-busy
        {...rest}
      >
        <div
          className={classnames(
            'dnb-h--xx-large',
            'dnb-skeleton',
            'dnb-skeleton--shape',
            'dnb-space__bottom--large'
          )}
          style={{
            width: '50%'
          }}
        >
          &zwnj;
        </div>

        {this.rowsLength.map((p, i) => (
          <div
            key={i}
            className={classnames(
              'dnb-p',
              'dnb-skeleton',
              'dnb-skeleton--shape',
              'dnb-space__top--x-small'
            )}
            style={{
              width: `${p}%`
            }}
          >
            &zwnj;
          </div>
        ))}

        {children}
      </div>
    )
  }
}
