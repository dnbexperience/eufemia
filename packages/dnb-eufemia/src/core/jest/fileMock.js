export default 'static file content'
