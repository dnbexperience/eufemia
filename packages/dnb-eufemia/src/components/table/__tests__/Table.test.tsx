import React from 'react'
import { render, screen } from '@testing-library/react'
import { loadScss, axeComponent } from '../../../core/jest/jestSetup'
import { BasicTable } from './TableMocks'
import Table, { TableAllProps } from '../Table'

const NODE_ENV = process.env.NODE_ENV
const log = globalThis.console.log

beforeEach(() => {
  window.IntersectionObserver = jest.fn().mockImplementation(() => {
    return {
      observe: jest.fn(),
      disconnect: jest.fn(),
    }
  })
})

afterEach(() => {
  process.env.NODE_ENV = NODE_ENV
  globalThis.console.log = log

  delete window.IntersectionObserver
})

describe('Table', () => {
  it('renders with props as an object', () => {
    const props: TableAllProps = { children: null }
    render(<Table {...props} />)

    expect(document.querySelector('.dnb-table')).toBeTruthy()
  })

  it('should contain basis HTML classes by default', () => {
    render(
      <Table>
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toEqual([
      'dnb-table',
      'dnb-table__variant--generic',
      'dnb-table__size--large',
    ])
  })

  it('should set variant', () => {
    render(
      <Table variant="generic">
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toEqual([
      'dnb-table',
      'dnb-table__variant--generic',
      'dnb-table__size--large',
    ])
  })

  it('should set size', () => {
    render(
      <Table size="medium">
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toEqual([
      'dnb-table',
      'dnb-table__variant--generic',
      'dnb-table__size--medium',
    ])
  })

  it('should include custom className', () => {
    render(
      <Table className="custom-class" aria-label="custom-label">
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toEqual(
      expect.arrayContaining([
        'dnb-table',
        'dnb-table__variant--generic',
        'dnb-table__size--large',
        'custom-class',
      ])
    )
  })

  it('should include custom HTML attributes', () => {
    render(
      <Table aria-label="custom-label">
        <BasicTable />
      </Table>
    )

    expect(screen.queryByRole('table').getAttribute('aria-label')).toEqual(
      'custom-label'
    )
  })

  it('should set the fixed class', () => {
    render(
      <Table fixed>
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toEqual(
      expect.arrayContaining([
        'dnb-table',
        'dnb-table__variant--generic',
        'dnb-table__size--large',
        'dnb-table--fixed',
      ])
    )
  })

  it('should set the border class', () => {
    render(
      <Table border>
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toContain(
      'dnb-table--border'
    )
  })

  it('should set the outline class', () => {
    render(
      <Table outline>
        <BasicTable />
      </Table>
    )

    expect(Array.from(screen.queryByRole('table').classList)).toContain(
      'dnb-table--outline'
    )
  })

  it('should support spacing props', () => {
    render(
      <Table top="2rem">
        <BasicTable />
      </Table>
    )

    const element = screen.queryByRole('table')
    const attributes = Array.from(element.attributes).map(
      (attr) => attr.name
    )

    expect(attributes).toEqual(['class'])
    expect(Array.from(element.classList)).toEqual(
      expect.arrayContaining([
        'dnb-table',
        'dnb-table__variant--generic',
        'dnb-table__size--large',
        'dnb-space__top--large',
      ])
    )
  })

  it('should support skeleton props', () => {
    render(
      <Table skeleton>
        <BasicTable />
      </Table>
    )

    const element = screen.queryByRole('table')
    const attributes = Array.from(element.attributes).map(
      (attr) => attr.name
    )

    expect(attributes).toEqual(['class'])
    expect(Array.from(element.classList)).toEqual(
      expect.arrayContaining([
        'dnb-table',
        'dnb-table__variant--generic',
        'dnb-table__size--large',
        'dnb-skeleton',
        'dnb-skeleton--font',
      ])
    )
  })
})

describe('Table aria', () => {
  it('should validate', async () => {
    const Component = render(
      <Table>
        <BasicTable />
      </Table>
    )
    expect(await axeComponent(Component)).toHaveNoViolations()
  })
})

describe('Table scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-table-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
