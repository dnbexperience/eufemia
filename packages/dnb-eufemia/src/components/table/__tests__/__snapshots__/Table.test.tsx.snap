// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table scss have to match default theme snapshot 1`] = `
"/*
* Table theme
*
*/
/**
 * This file is only used to make themes independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
.dnb-table > thead > tr > th::after,
.dnb-table > tbody > tr:last-of-type > td::after,
.dnb-table > thead > .dnb-table__tr > .dnb-table__th::after,
.dnb-table > tbody > .dnb-table__tr:last-of-type > .dnb-table__td::after {
  content: '';
  position: absolute;
  left: 0;
  right: 0;
  bottom: -1px;
  height: 1px;
  background-color: var(--color-black-8); }
"
`;

exports[`Table scss have to match snapshot 1`] = `
"/*
* DNB Table
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
.dnb-table {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-basis);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * Ensure consistency and use the same as HTML reset -> html {...}
  * between base and code package
  */
  -moz-tab-size: 4;
  tab-size: 4;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  word-break: break-word;
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-table *,
  .dnb-table ::before,
  .dnb-table ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-table ::before,
  .dnb-table ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
* Table component
*
*/
/*
* Button mixins
*
*/
/*
* Table component
*
*/
.dnb-table > tr > th,
.dnb-table > thead > tr > th, .dnb-table__th {
  padding: var(--spacing-x-large) var(--spacing-small) var(--spacing-x-small);
  font-weight: var(--font-weight-medium);
  color: var(--color-black);
  vertical-align: bottom;
  background-color: var(--color-white); }
  .dnb-table > tr > th:not([align]),
  .dnb-table > thead > tr > th:not([align]), .dnb-table__th:not([align]) {
    text-align: inherit; }
  @supports (-moz-appearance: meterbar) and (background-blend-mode: difference, normal) {
    .dnb-table > tr > th,
    .dnb-table > thead > tr > th, .dnb-table__th {
      padding-bottom: calc(var(--spacing-x-small) - 0.5px); } }

.dnb-table__th__horizontal {
  display: flex;
  align-items: flex-end; }

/*
* Table component
*
*/
.dnb-table > tr > td,
.dnb-table > tbody > tr > td, .dnb-table__td,
.dnb-table > tr > th[scope='row'] {
  padding: var(--spacing-small);
  padding-top: calc(var(--spacing-small) * 1.25);
  padding-bottom: calc(var(--spacing-small) * 1.25);
  color: var(--theme-color-black-80, currentColor);
  vertical-align: baseline; }

.dnb-table__td--no-spacing,
.dnb-table td.dnb-table__td--no-spacing {
  padding: 0; }

/*
* Table component
*
*/
.dnb-table > tbody > tr,
.dnb-table .dnb-table__tr,
.dnb-table .dnb-table__tr--even {
  background-color: var(--color-lavender); }

.dnb-table > tbody > tr:not(.dnb-table__tr--even):nth-of-type(2n),
.dnb-table .dnb-table__tr:not(.dnb-table__tr--even):nth-of-type(2n),
.dnb-table .dnb-table__tr--odd {
  background-color: var(--color-white); }

/*
* Table component
*
*/
.dnb-table > tr > th,
.dnb-table > tr > td,
.dnb-table > thead > tr > th,
.dnb-table > tbody > tr > td, .dnb-table__th, .dnb-table__td {
  position: relative;
  font-size: var(--font-size-basis);
  line-height: var(--line-height-basis);
  border-spacing: 0;
  word-break: keep-all; }

/*
* Table component
*
*/
.dnb-table__sticky-helper > td {
  display: block;
  overflow: hidden;
  padding: 0 !important;
  height: 0; }

.dnb-table tr.sticky th {
  position: sticky;
  top: var(--table-top, 0);
  z-index: 2; }
  .dnb-table tr.sticky th::before {
    content: '';
    position: absolute;
    opacity: 0;
    left: 0;
    right: 0;
    bottom: 0;
    height: 6px;
    box-shadow: var(--shadow-default-x) -2px 12px 8px var(--shadow-default-color);
    clip-path: inset(6px 0 -48px 0);
    transition: opacity 300ms ease-out; }
    [data-visual-test-wrapper] .dnb-table tr.sticky th::before {
      transition: none; }

.dnb-table tr.sticky.show-shadow th::before {
  opacity: 1; }

.dnb-table {
  display: table;
  table-layout: auto;
  width: 100%;
  overflow: auto;
  margin-top: 0;
  margin-bottom: var(--spacing-small);
  border-spacing: 0;
  border-collapse: collapse;
  /* stylelint-disable */
  /* stylelint-enable */ }
  .dnb-table--fixed {
    table-layout: fixed;
    width: auto; }
  .dnb-table--no-wrap {
    white-space: nowrap; }
  .dnb-table, .dnb-table--left {
    text-align: left; }
  .dnb-table--right {
    text-align: right; }
  .dnb-table--center {
    text-align: center; }
  .dnb-table > caption {
    font-size: var(--font-size-small);
    margin-top: var(--spacing-x-small);
    caption-side: bottom; }
  .dnb-table--small,
  .dnb-table--small > tr > th,
  .dnb-table--small > tr > td,
  .dnb-table--small > thead > tr > th,
  .dnb-table--small > tbody > tr > td {
    font-size: var(--font-size-small); }
  .dnb-table > tr.dnb-table--small > th,
  .dnb-table > tr.dnb-table--small > td,
  .dnb-table > thead > tr.dnb-table--small > th,
  .dnb-table > tbody > tr.dnb-table--small > td {
    font-size: var(--font-size-small); }
  .dnb-table > tr > th.dnb-table--small,
  .dnb-table > tr > td.dnb-table--small,
  .dnb-table > thead > tr > th.dnb-table--small,
  .dnb-table > tbody > tr > td.dnb-table--small, .dnb-table__th.dnb-table--small, .dnb-table__td.dnb-table--small {
    font-size: var(--font-size-small); }
  .dnb-table--x-small,
  .dnb-table--x-small > tr > th,
  .dnb-table--x-small > tr > td,
  .dnb-table--x-small > thead > tr > th,
  .dnb-table--x-small > tbody > tr > td {
    font-size: var(--font-size-x-small); }
  .dnb-table > tr.dnb-table--x-small > th,
  .dnb-table > tr.dnb-table--x-small > td,
  .dnb-table > thead > tr.dnb-table--x-small > th,
  .dnb-table > tbody > tr.dnb-table--x-small > td {
    font-size: var(--font-size-x-small); }
  .dnb-table > tr > th.dnb-table--x-small,
  .dnb-table > tr > td.dnb-table--x-small,
  .dnb-table > thead > tr > th.dnb-table--x-small,
  .dnb-table > tbody > tr > td.dnb-table--x-small, .dnb-table__th.dnb-table--x-small, .dnb-table__td.dnb-table--x-small {
    font-size: var(--font-size-x-small); }
  .dnb-table.dnb-skeleton > * {
    -webkit-text-fill-color: var(--skeleton-color); }
  .dnb-table > thead > tr > th.dnb-table--sortable,
  .dnb-table .dnb-table__th.dnb-table--sortable {
    color: var(--color-emerald-green); }
    .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button,
    .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button {
      position: relative;
      z-index: 1;
      align-items: flex-end;
      box-shadow: none;
      line-height: var(--line-height-basis);
      margin: 0;
      padding: 0;
      padding-right: 0.5rem;
      color: inherit;
      font-weight: var(--font-weight-medium);
      text-align: inherit;
      font-size: inherit; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button > .dnb-icon,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button > .dnb-icon {
        opacity: 0;
        transition: opacity 200ms ease-out, transform 500ms ease-out; }
        [data-visual-test-wrapper] .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button > .dnb-icon, [data-visual-test-wrapper]
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button > .dnb-icon {
          transition: none; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__text,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__text {
        margin: 0;
        font-size: inherit; }
        .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__text::after,
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__text::after {
          right: 1rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__icon,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button .dnb-button__icon {
        align-self: flex-end;
        margin-top: auto;
        margin-bottom: 0.25rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:not(:focus) .dnb-button__text::after,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:not(:focus) .dnb-button__text::after {
        opacity: 1;
        color: var(--color-emerald-green); }
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover[disabled], html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover[disabled] {
        cursor: not-allowed; }
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]), html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]) {
        color: var(--color-sea-green); }
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]):not(:focus) .dnb-icon, html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]):not(:focus) .dnb-icon {
          opacity: 1; }
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]) .dnb-button__text::after, html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:not([disabled]) .dnb-button__text::after {
          opacity: 0; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus[disabled],
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus[disabled],
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus[disabled],
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus[disabled] {
        cursor: not-allowed; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before {
        content: '';
        position: absolute;
        top: -0.5rem;
        bottom: -0.5rem;
        left: -1rem;
        right: 0.5rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before,
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled])::before {
        right: -0.5rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled]) .dnb-icon,
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled]) .dnb-icon,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled]) .dnb-icon,
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:focus:not([disabled]) .dnb-icon {
        opacity: 1; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active[disabled],
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active[disabled],
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active[disabled],
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active[disabled] {
        cursor: not-allowed; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled]),
      html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled]),
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled]),
      html:not([data-whatintent='touch'])
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled]) {
        color: var(--color-emerald-green); }
        .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
          content: '';
          position: absolute;
          z-index: 1;
          top: 0;
          left: -0.5rem;
          bottom: 0;
          right: -0.5rem;
          height: inherit;
          border-radius: inherit;
          outline: none; }
          html[data-whatinput='mouse'] .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
          html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
          .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
          html:not([data-whatintent='touch'])
          .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
            --border-color: var(--color-emerald-green);
            box-shadow: 0 0 0 0.125rem var(--border-color);
            border-color: transparent; }
            @media screen and (-ms-high-contrast: none) {
              html[data-whatinput='mouse'] .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
              html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
              .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='mouse']
              html:not([data-whatintent='touch'])
              .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
                box-shadow: 0 0 0 0.125rem var(--color-emerald-green); } }
        .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
          content: '';
          position: absolute;
          z-index: 1;
          top: 0;
          left: -0.5rem;
          bottom: 0;
          right: -0.5rem;
          height: inherit;
          border-radius: inherit;
          outline: none; }
          html[data-whatinput='touch'] .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
          html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
          .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
          html:not([data-whatintent='touch'])
          .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
            --border-color: var(--color-emerald-green);
            box-shadow: 0 0 0 0.125rem var(--border-color);
            border-color: transparent; }
            @media screen and (-ms-high-contrast: none) {
              html[data-whatinput='touch'] .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
              html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
              .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before, html[data-whatinput='touch']
              html:not([data-whatintent='touch'])
              .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
                box-shadow: 0 0 0 0.125rem var(--color-emerald-green); } }
        .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
          content: '';
          position: absolute;
          top: -0.5rem;
          bottom: -0.5rem;
          left: -1rem;
          right: 0.5rem; }
        .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before,
        html:not([data-whatintent='touch'])
        .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:active:not([disabled])::before {
          right: -0.5rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:focus:not(:active) .dnb-button__text::after,
      .dnb-table .dnb-table__th.dnb-table--sortable .dnb-table__sort-button.dnb-button:hover:focus:not(:active) .dnb-button__text::after {
        visibility: hidden; }
    .dnb-table > thead > tr > th.dnb-table--sortable[align='right'],
    .dnb-table .dnb-table__th.dnb-table--sortable[align='right'] {
      padding-right: 0.5rem; }
      .dnb-table > thead > tr > th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button,
      .dnb-table .dnb-table__th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button {
        flex-direction: row-reverse; }
        .dnb-table > thead > tr > th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right,
        .dnb-table .dnb-table__th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right {
          padding-right: 0.5rem; }
          .dnb-table > thead > tr > th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right .dnb-button__text,
          .dnb-table .dnb-table__th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right .dnb-button__text {
            padding-right: 0; }
            .dnb-table > thead > tr > th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right .dnb-button__text::after,
            .dnb-table .dnb-table__th.dnb-table--sortable[align='right'] .dnb-table__sort-button.dnb-button--tertiary.dnb-button--icon-position-right .dnb-button__text::after {
              right: 0; }
  .dnb-table > thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button:not(:hover) .dnb-button__text::after,
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button:not(:hover) .dnb-button__text::after {
    opacity: 0; }
  .dnb-table > thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button .dnb-icon,
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button .dnb-icon {
    opacity: 1; }
  html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button:hover[disabled], html:not([data-whatintent='touch'])
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button:hover[disabled] {
    cursor: not-allowed; }
  html:not([data-whatintent='touch']) .dnb-table > thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:not([disabled]) .dnb-button__text::after, html:not([data-whatintent='touch'])
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:not([disabled]) .dnb-button__text::after {
    color: var(--color-sea-green);
    opacity: 1; }
  .dnb-table > thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:focus:not(:active) .dnb-button__text::after,
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:focus:not(:active) .dnb-button__text::after {
    visibility: visible; }
  html[data-whatinput='keyboard']
.dnb-table >
thead > tr > th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:focus .dnb-button__text::after, html[data-whatinput='keyboard']
  .dnb-table .dnb-table__th.dnb-table--active .dnb-table__sort-button.dnb-button:hover:focus .dnb-button__text::after {
    visibility: hidden; }
  .dnb-table > thead > tr > th.dnb-table--reversed .dnb-table__sort-button.dnb-button .dnb-icon,
  .dnb-table .dnb-table__th.dnb-table--reversed .dnb-table__sort-button.dnb-button .dnb-icon {
    transform: rotate(180deg);
    transform-origin: 50% 50%; }
  .dnb-table .dnb-table__th .dnb-table__help-button,
  .dnb-table > thead > tr > th .dnb-table__help-button {
    margin-left: 0.5rem; }

.dnb-spacing .dnb-table p,
.dnb-spacing .dnb-table ul,
.dnb-spacing .dnb-table ol,
.dnb-spacing .dnb-table dl,
.dnb-spacing .dnb-table pre {
  margin-top: 0;
  margin-bottom: 0; }
"
`;
